### ISSUE TITLE
[API Engine] To develop command 5; (Broker configuration)

### ISSUE THREAD

**Issue by ralcaide** [Sinapse-Energia/EC-M2M#21]
It is necessary to develop the command 5; (Broker configuration) for the beta version. The test devices for Vodafone should implement this command

Description: Changes server configuration parameters
Format: ‘5;URI;PORT;USR,PASS’

Parameters: 
URI: server IP address or URL
PORT: server port address.
USR: User
PASS: Password

**Comment by juanra-bascones** [comment]
Two questions about this.
1) Where these values have to be stored?
Do they have to persist in Flash Memory o other non volatile storage?
 Otherwise, the new values will be lost alter restart, wdt or power faliure
2) Have this command to trigger an inmediate reconnection with the new values, or they have only to be stored for later use, just in case?

**Comment by ralcaide** [comment]
Responding your questions:

1. Yes, it should be saved in a non volatile storage. @francisjjp, do you have already the memory map for this developed?. @juanra-bascones , by the moment, could you save it in some part of the Flash?. A question: The Context variables are lost after a restart? If not, the Broker could be a context variable. If yes, we need to think about that because some variables should be saved in a non volatile storage like the energy consumed.

2. This command should trigger an inmmediate reconnection to the new broker. But this is dangerous if the connection is not performed. If some problem occurs during the first reconnection it should connect again to the precious Broker (as a security mechanism)

**Comment by juanra-bascones** [comment]
I would like someone to give a short example about how to store (and retrieve) data form de flash memory.
As probably there have to be much many things to persist, maybe could be clever to foresee the persistence as an optional attribute of the context items. 
In relation with the other point (keeping the previous broker), we have to define the behaviour in each scenario:
1) If the connection succeeds, the new broker parameters should be used hereinafter, the previous would be forgot, and the new ones become default   It is correct?
2) If the connection fails, it will try to reconnect back to the old one, and will forget about the new one, and the former remain as default.  It is correct?

**Comment by juanra-bascones** [comment]
In order to make better testing of this command, (using both, the cloud and de sinapse brokers), would be very advisable  update the sinapse AWS mosquitto server to a more recent version, supporting mqttversion 4.
This will  avoid to have to give rigth now an ultimate decission to the issue of how to handle de multiple mqttversion support form de EC side.

**Comment by ralcaide** [comment]
Yes, @juanra-bascones you are totally right in relation saving the needed vars in flash memory. @francisjjp  will make this work. You can focus only in the implementation of the API message & protocol and by the moment saving the content in a volatile memory (as a normal context variable)

In relation to the scenario:
1. YES
2. YES

**Comment by juanra-bascones** [comment]
I have a first version of the command in external-master.
Is a very basic versión, without persistence and without any contingence mechanism in case of failure with the new broker parameters, but it works well with those limitations.
I', going to push it to the cloud inmediately

**Comment by juanra-bascones** [comment]
I have been thinking the best approach to the contingence o¡in case of failure to the new broker provided by command 5;
The point is to have TWO sets of broker parameters:  ONE, let´s call it the "configured parameters", and the OTHER, that we can call "last succesfull connection".
The first one comes initially from the config, and changes each time a command 5;  is received.
The second one is automatically updated upon succesfull broker connection.
Both are stored in the Context and both deserve to persist.
The connection process, wherever, first takes the first ONE.
If succeds, updates the second with the used parameters.
If fails, keeps the first contens, but retries to connect wit the second one.

**Comment by ralcaide** [comment]
@juanra-bascones , I really like your proposed mechanism. It is safe. Please, implement it. We lack only the persistence.

**Comment by ralcaide** [comment]
@juanra-bascones , so this issue is almost done but without persistence, isn't?. Waiting for #22 in order to finish the development of this issue

**Comment by juanra-bascones** [comment]
Made the integration with persistence.
Tested in several scenarios (connection failures, etc...) in order to verify the regular broker change, and also the contingency mechanism.
Times involved are quite a bite high, but behaviour is consistent with expectations.
